lib/process-services-cloud/src/lib/form/services/content-cloud-node-selector.service.spec.ts

```diff
@@ -0,0 +1,124 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { AlfrescoApiService, NotificationService, setupTestBed } from '@alfresco/adf-core';
+import { ProcessServiceCloudTestingModule } from '../../testing/process-service-cloud.testing.module';
+import { TranslateModule } from '@ngx-translate/core';
+import { ContentCloudNodeSelectorService } from 'process-services-cloud';
+import { MatDialog, MatDialogModule } from '@angular/material/dialog';
+import { of, Subject } from 'rxjs';
+
+describe('ContentCloudNodeSelectorService', () => {
+    let service: ContentCloudNodeSelectorService;
+    let apiService: AlfrescoApiService;
+    let notificationService: NotificationService;
+    let getNodeSpy: jasmine.Spy;
+    let dialog: MatDialog;
+    let openDialogSpy: jasmine.Spy;
+    let showWarningSpy: jasmine.Spy;
+
+    const relativePathNodeResponseBody = {
+        entry: {
+            id: 'mock-relative-path-node-id'
+        }
+    };
+
+    const aliasNodeResponseBody = {
+        entry: {
+            id: 'mock-alias-node-id'
+        }
+    };
+
+    setupTestBed({
+        imports: [TranslateModule.forRoot(), ProcessServiceCloudTestingModule, MatDialogModule]
+    });
+
+    beforeEach(() => {
+        service = TestBed.inject(ContentCloudNodeSelectorService);
+        notificationService = TestBed.inject(NotificationService);
+        apiService = TestBed.inject(AlfrescoApiService);
+        dialog = TestBed.inject(MatDialog);
+
+        showWarningSpy = spyOn(notificationService, 'showWarning');
+        openDialogSpy = spyOn(dialog, 'open').and.returnValue({
+            afterOpened: () => of({}),
+            afterClosed: () => of({}),
+            componentInstance: {
+                body: '',
+                error: new Subject<any>()
+            }
+        });
+        getNodeSpy = spyOn(apiService.nodesApi, 'getNode');
+    });
+
+    it('should be able to set sourceNodeNotFound value to true if the relative path is invalid/deleted', async () => {
+        expect(service.sourceNodeNotFound).toBe(false);
+
+        getNodeSpy.and.returnValue(Promise.reject('Not exists'));
+        await service.fetchNodeIdFromRelativePath('mock-alias', { relativePath: 'mock-wrong-relativePath' });
+
+        expect(getNodeSpy).toHaveBeenCalledWith('mock-alias', {
+            relativePath: 'mock-wrong-relativePath'
+        });
+        expect(service.sourceNodeNotFound).toBe(true);
+    });
+
+    it('should be able to set sourceNodeNotFound value to false after the dialog close', async () => {
+        service.sourceNodeNotFound = true;
+        service.openUploadFileDialog('nodeId', 'single', true, true);
+
+        expect(openDialogSpy).toHaveBeenCalled();
```

Comment:
It is better not to add extra checks along the way of testing a particular thing. A unit test should test one thing. Here you called the `service.openUploadFileDialog` and you check that the dialog is opened when your test title says something else. Then you test that the sourceNodeNotFound should be false. 

I see 2 tests here (If opening the dialog is not covered already, it probably is). This causes a small refactoring to require a lot of changes when there is no true benefit for adding this check. If there is no test about that then there should be:
- 1 test that checks that the dialog opens.
- 1 test (your test) that should only check what the title says. `should be able to set sourceNodeNotFound value to false after the dialog close`

Reply:
I see this spec looks good. This spec is testing enter flow once you click on the widget.

1. Is Dialog opened
2. Close the dialog
3. Update sourceNodeNotFound flag after dialog close. (Without the first and second steps, we won't reach 3rd step right.)


It would be always better to check each step (like we do on e2e) instead of testing a single check. I can remove it but I do not see any benefit from it.

Reply:
1. You open the dialog
2. You close it
3. you check what you need to check
This is correct I am not saying it is wrong.

The thing that I think you should not do is check that the dialog is opened.
Right now the flow is:

1. You open the dialog
2. You check that the dialog is open (That is the one I am suggesting to remove)
3. You close it
4. you check what you need to check

To respect that each test should test one thing you should have 2 tests to cover them
Test 1) Open the dialog -> Check that the dialog is opened
Test 2) open the dialog -> Close the dialog -> Check that the sourceNodeNotFound is false

Reply:
Ok Removed it 

Here is the separate test for Open the dialog

https://github.com/Alfresco/alfresco-ng2-components/pull/7173/files#diff-b9940e991191204dd31c27360bea21130ae12a0c32a376a75863c6e75a0ad5ccR71-R90